src/cljam/io/bam/core.clj

```diff
@@ -1,36 +1,44 @@
 (ns cljam.io.bam.core
   "The core of BAM features."
   (:require [clojure.java.io :as cio]
+            [clojure.string :as cstr]
             [cljam.io.bam [common :refer [bam-magic]]
                           [reader :as reader]
                           [writer :as writer]]
             [cljam.io.bam-index :as bai]
             [cljam.io.util.lsb :as lsb])
   (:import java.util.Arrays
-           [java.io DataInputStream DataOutputStream IOException]
+           [java.io DataInputStream DataOutputStream IOException FileNotFoundException]
            [bgzf4j BGZFInputStream BGZFOutputStream]
            cljam.io.bam.reader.BAMReader
            cljam.io.bam.writer.BAMWriter))
 
 ;; Reading
 ;; -------
 
-(defn- bam-index [f & {:keys [ignore]
-                       :or {ignore false}}]
-  (if-not ignore
-    (let [bai-f (str f ".bai")]
-      (if (.isFile (cio/file bai-f))
-        (bai/bam-index bai-f)
-        (throw (IOException. "Could not find BAM Index file"))))))
+(defn- bam-index
+  "Load an index file (BAI) for the given BAM file."
+  [bam-path & [{:keys [ignore]}]]
+  (let [bai-path (->> ["$1.bai" ".bai" "$1.BAI" ".BAI"]
+                      (eduction
+                       (comp
+                        (map #(cstr/replace bam-path #"(?i)(\.bam)" %))
+                        (filter #(.isFile (cio/file %)))))
+                      first)]
+    (if (and bai-path (not ignore))
```

Comment:
Thanks for your comment.
As you said, I've changed the behavior from the master version in this PR.

The `:ignore` option was originally introduced to skip loading an index file before constructing BAMReader.
But because `cljam.io.bam.core/bam-index` is now always called in `delay`,
the actual loading of the index file (or throwing an exception) will happen when the first time a random access is queried.
So it's just a matter of the position of throwing exceptions:
- master: nil-handling part in BAM reader
- this PR: inside of the bam-index function

And I think the latter is simpler and more consistent.